integration/traffic/historicalspeed/historical_speeds.go

```diff
@@ -0,0 +1,55 @@
+package historicalspeed
+
+import "time"
+
+// Speeds stores historical speeds.
+type Speeds struct {
+	dailyPatterns       map[uint32]dailyPattern
+	way2PatternsMapping map[int64]*mappingItem // indexed by wayID: positive means forward, negative means backward
+
+	// allow multiple files
+	dailyPatternsFilePath        []string
+	ways2PatternsMappingFilePath []string
+}
+
+// New create a empty Speeds object.
+func New(dailyPatternsFilePath, ways2PatternsMappingFilePath []string) *Speeds {
+
+	return &Speeds{
+		dailyPatterns:       map[uint32]dailyPattern{},
+		way2PatternsMapping: map[int64]*mappingItem{},
+
+		dailyPatternsFilePath:        dailyPatternsFilePath,
+		ways2PatternsMappingFilePath: ways2PatternsMappingFilePath,
+	}
+}
+
+// Load loads contents from file into memory.
+func (s *Speeds) Load() error {
+	if err := s.loadDailyPatterns(); err != nil {
+		return err
+	}
+	if err := s.loadWaysPatternsMapping(); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// QueryHistoricalSpeed return the speed for a way at a specified time
+// wayID: positive means forward, negative means backward
+// t: UTC time
+func (s *Speeds) QueryHistoricalSpeed(wayID int64, t time.Time) float64 {
+	//TODO:
+	return 0.0
+}
```

Comment:
Enable historic speed is a very important feature for OSRM, expect for this great feature.

Just for discussion:
1. So those data will be loaded into memory, since historical data already be pre-processed for embedded use, I think this should be fine.  Just confirm
2. There is another table you need: a mapping between <wayid, historical pattern id>
  May be we need a micro service for our team: not only get by <wayid> or <wayids>, but retrieve data data for all of the data for specific pattern, and most importantly, it should be able to cache result of query.

Reply:
1. I didn't use the data processed for embedded. After dig into historical speed technologies, I prefer to use the data before process for embedded. It should be fine enough on cloud side. I'll share you more internal information. 
2. I think it's a good idea. We can discuss it later to build a centralize micro service once we have multiple components need to retrieve historical speed.